collector.c

```diff
@@ -5,301 +5,221 @@
  * Copyright (c) 2015-2016, Postgres Professional
  *
  * IDENTIFICATION
- *	  contrib/pg_wait_sampling/pg_wait_sampling.c
+ *	  contrib/pg_wait_sampling/collector.c
  */
 #include "postgres.h"
 
-#include "catalog/pg_type.h"
 #if PG_VERSION_NUM >= 130000
 #include "common/hashfn.h"
 #endif
-#include "funcapi.h"
-#include "miscadmin.h"
+#include "pgstat.h"
 #include "postmaster/bgworker.h"
+#if PG_VERSION_NUM >= 130000
+#include "postmaster/interrupt.h"
+#endif
 #include "storage/ipc.h"
-#include "storage/procarray.h"
+#include "storage/proc.h"
 #include "storage/procsignal.h"
-#include "storage/shm_mq.h"
-#include "storage/shm_toc.h"
-#include "storage/spin.h"
-#include "utils/memutils.h"
-#include "utils/resowner.h"
-#include "pgstat.h"
+#include "utils/guc.h"
 
 #include "compat.h"
 #include "pg_wait_sampling.h"
 
+static const double USAGE_INIT = 1.0;
+static const double USAGE_INCREASE = 1.0;
+static const double USAGE_DECREASE_FACTOR = 0.99;
```

Comment:
I don't have an intuition for these values. If I my understanding of the code is correct, an entry with count = 10 will have usage = 10, if there were no eviction cycles. Then we need more than 200 eviction cycles for the usage of the entry to drop back to 1. It seems that old entries are never evicted.

Okay, after some testing I see that my reasoning is wrong. I'll try to figure this out.

Reply:
I did some more testing and I think there is a problem with old entries not being evicted. Let me describe what I observe.

To make the output smaller I set MaxProfileEntries=20 and USAGE_DEALLOC_MIN_NUM=3. I also set ProfilePeriod=100 (ms).

After starting Postgres and connecting with psql I see:
```
postgres=# select * from pg_wait_sampling_profile order by pid desc;
  pid  | event_type |        event        | queryid | count 
-------+------------+---------------------+---------+-------
 65078 | Client     | ClientRead          |       0 |    78
 65077 | Activity   | LogicalLauncherMain |       0 |   106
 65075 | Activity   | AutoVacuumMain      |       0 |   106
 65074 | Activity   | WalWriterMain       |       0 |   106
 65072 | Activity   | BgWriterMain        |       0 |   106
 65071 | Activity   | CheckpointerMain    |       0 |   106
(6 rows)
```
That's fine. Next I create 10 new sessions with pgpench:
```
pgbench -n -c10 -t1 -f <(echo "select pg_sleep(1)")
```

And the profile becomes:
```
postgres=# select * from pg_wait_sampling_profile order by pid desc;
  pid  | event_type |        event        |      queryid       | count 
-------+------------+---------------------+--------------------+-------
 65092 | Timeout    | PgSleep             | 440101247839410938 |    10
 65091 | Timeout    | PgSleep             | 440101247839410938 |    10
 65090 | Timeout    | PgSleep             | 440101247839410938 |    10
 65089 | Timeout    | PgSleep             | 440101247839410938 |    10
 65088 | Timeout    | PgSleep             | 440101247839410938 |    10
 65087 | Timeout    | PgSleep             | 440101247839410938 |    10
 65086 | Timeout    | PgSleep             | 440101247839410938 |    10
 65085 | Timeout    | PgSleep             | 440101247839410938 |    10
 65084 | Timeout    | PgSleep             | 440101247839410938 |    10
 65083 | Timeout    | PgSleep             | 440101247839410938 |    10
 65078 | Client     | ClientRead          |                  0 |   221
 65077 | Activity   | LogicalLauncherMain |                  0 |   249
 65075 | Activity   | AutoVacuumMain      |                  0 |   249
 65074 | Activity   | WalWriterMain       |                  0 |   249
 65072 | Activity   | BgWriterMain        |                  0 |   249
 65071 | Activity   | CheckpointerMain    |                  0 |   249
(16 rows)
```
That's also fine and expected.

Now, to test eviction, I run pgbench for the second time and see:
```
postgres=# select * from pg_wait_sampling_profile order by pid desc;
  pid  | event_type |        event        |      queryid       | count 
-------+------------+---------------------+--------------------+-------
 65107 | Timeout    | PgSleep             | 440101247839410938 |     1
 65106 | Timeout    | PgSleep             | 440101247839410938 |     1
 65105 | Timeout    | PgSleep             | 440101247839410938 |     1
 65101 | Timeout    | PgSleep             | 440101247839410938 |     1
 65092 | Timeout    | PgSleep             | 440101247839410938 |    10
 65091 | Timeout    | PgSleep             | 440101247839410938 |    10
 65090 | Timeout    | PgSleep             | 440101247839410938 |    10
 65089 | Timeout    | PgSleep             | 440101247839410938 |    10
 65088 | Timeout    | PgSleep             | 440101247839410938 |    10
 65087 | Timeout    | PgSleep             | 440101247839410938 |    10
 65086 | Timeout    | PgSleep             | 440101247839410938 |    10
 65085 | Timeout    | PgSleep             | 440101247839410938 |    10
 65084 | Timeout    | PgSleep             | 440101247839410938 |    10
 65083 | Timeout    | PgSleep             | 440101247839410938 |    10
 65078 | Client     | ClientRead          |                  0 |   444
 65077 | Activity   | LogicalLauncherMain |                  0 |   472
 65075 | Activity   | AutoVacuumMain      |                  0 |   472
 65074 | Activity   | WalWriterMain       |                  0 |   472
 65072 | Activity   | BgWriterMain        |                  0 |   472
 65071 | Activity   | CheckpointerMain    |                  0 |   472
(20 rows)
```
There are only 4 new entries, old entries left after the previous run are not evicted. After running pgbench two more times:
```
postgres=# select * from pg_wait_sampling_profile order by pid desc;
  pid  | event_type |        event        |      queryid       | count 
-------+------------+---------------------+--------------------+-------
 65130 | Timeout    | PgSleep             | 440101247839410938 |     1
 65126 | Timeout    | PgSleep             | 440101247839410938 |     1
 65092 | Timeout    | PgSleep             | 440101247839410938 |    10
 65091 | Timeout    | PgSleep             | 440101247839410938 |    10
 65090 | Timeout    | PgSleep             | 440101247839410938 |    10
 65089 | Timeout    | PgSleep             | 440101247839410938 |    10
 65088 | Timeout    | PgSleep             | 440101247839410938 |    10
 65087 | Timeout    | PgSleep             | 440101247839410938 |    10
 65086 | Timeout    | PgSleep             | 440101247839410938 |    10
 65085 | Timeout    | PgSleep             | 440101247839410938 |    10
 65084 | Timeout    | PgSleep             | 440101247839410938 |    10
 65083 | Timeout    | PgSleep             | 440101247839410938 |    10
 65078 | Client     | ClientRead          |                  0 |  1366
 65077 | Activity   | LogicalLauncherMain |                  0 |  1394
 65075 | Activity   | AutoVacuumMain      |                  0 |  1394
 65074 | Activity   | WalWriterMain       |                  0 |  1394
 65072 | Activity   | BgWriterMain        |                  0 |  1211
 65072 | Activity   | BgWriterHibernate   |                  0 |   182
 65071 | Activity   | CheckpointerMain    |                  0 |  1394
(19 rows)

postgres=# select * from pg_wait_sampling_profile order by pid desc;
  pid  | event_type |        event        |      queryid       | count 
-------+------------+---------------------+--------------------+-------
 65149 | Timeout    | PgSleep             | 440101247839410938 |     1
 65148 | Timeout    | PgSleep             | 440101247839410938 |     1
 65144 | Timeout    | PgSleep             | 440101247839410938 |     1
 65092 | Timeout    | PgSleep             | 440101247839410938 |    10
 65091 | Timeout    | PgSleep             | 440101247839410938 |    10
 65090 | Timeout    | PgSleep             | 440101247839410938 |    10
 65089 | Timeout    | PgSleep             | 440101247839410938 |    10
 65088 | Timeout    | PgSleep             | 440101247839410938 |    10
 65087 | Timeout    | PgSleep             | 440101247839410938 |    10
 65086 | Timeout    | PgSleep             | 440101247839410938 |    10
 65085 | Timeout    | PgSleep             | 440101247839410938 |    10
 65084 | Timeout    | PgSleep             | 440101247839410938 |    10
 65083 | Timeout    | PgSleep             | 440101247839410938 |    10
 65078 | Client     | ClientRead          |                  0 |  1802
 65077 | Activity   | LogicalLauncherMain |                  0 |  1830
 65075 | Activity   | AutoVacuumMain      |                  0 |  1830
 65074 | Activity   | WalWriterMain       |                  0 |  1830
 65072 | Activity   | BgWriterMain        |                  0 |  1221
 65072 | Activity   | BgWriterHibernate   |                  0 |   608
 65071 | Activity   | CheckpointerMain    |                  0 |  1830
(20 rows)
```

Old entries are not evicted yet while their usage values dropped from 10 to 3.70.